### PR TITLE
[Setup] 데이터베이스 스키마 설계 및 연동

### LINKED ISSUE #9: [Feature] Supabase 초기 설정 및 연동
## 세부 기능

프로젝트에 사용할 Supabase 초기 설정을 진행하고, 연동합니다.

- Supabase 프로젝트 생성
- 환경 변수 설정
- 인증 관련 설정

## 완료 조건

- [x] 프리뷰의 데이터베이스 연결 성공

### LINKED ISSUE #10: [Feature] 블로그 데이터베이스 스키마 설계 및 구현
## 세부 기능

개인 블로그에 필요한 데이터베이스 테이블 구조를 설계하고 구현합니다.

필요한 테이블은 다음과 같습니다.
- posts: 블로그 포스트
- categories: 개발/학습 등 최상위 분류
- series: velog 시리즈 기능
- tags: 키워드 태그

## 완료 조건

- [x] 테이블 설정, 초기 더미 데이터 생성
- [x] RLS 정책 설정
- [x] 인덱스 설정

## 참고 자료

[//]: # '참고할 문서, 디자인, API 문서 등을 첨부해주세요.'

- [Supabase - Database](https://supabase.com/docs/guides/database/overview)
- [How can I run MDX queries against a PostgreSQL database?](https://stackoverflow.com/questions/28664483/how-can-i-run-mdx-queries-against-a-postgresql-database)

### PR DESCRIPTION
**Pull request by PMtHk**
## PR 설명

Supabase 내 데이터베이스를 설계하고, 연동을 진행했습니다.

## 작업 사항

- 환경변수를 관리하기 위해 zod 를 활용해 환경변수 검증을 진행했습니다.
- supabase 초기설정을 진행한 후 데이터베이스 스키마를 설정했습니다.
- supabase client 를 생성하는 로직을 작성했습니다.

연동 테스트는 진행했고, 해당 변경사항은 반영하지 않았습니다.

## 관련 이슈

해결하려는 이슈 번호를 연결해주세요.

- Resolved: #9, #10 

## 기타 이슈

- supabase 의 createBrowserClient 는 문제가 없었으나, createServerClient 는 cookie 관련 deprecated 된 속성들이 있어 별도의 설정이 필요했습니다.

### CODE REVIEW

File: src/libs/env/index.ts [hunk 1]
```diff
@@ -0,0 +1,3 @@
+export * from './schemas'
+export * from './public'
+export * from './server'
```
Comment (PMtHk) [inline src/libs/env/index.ts:3]: Overall, the codebase looks well-organized and readable. The file structure is clear, and the naming conventions are consistent. However, there are a few areas where improvements can be made:

1. Code organization & readability:
   - It's great that you are using separate files for different modules, but it would be helpful to provide more descriptive names for the files. For example, instead of just 'schemas', consider naming it 'userSchemas' or 'productSchemas' to make it clear what the file contains.
   - Consider breaking down larger files into smaller, more manageable chunks. This can improve readability and make it easier to maintain the code in the long run.

2. Basic security & error handling:
   - It's important to implement proper error handling mechanisms to prevent unexpected crashes and vulnerabilities. Make sure to validate user inputs, sanitize data, and handle exceptions gracefully.
   - Consider implementing input validation using a library like Joi or express-validator to ensure that only valid data is processed by the application.

3. Performance considerations:
   - While the codebase seems to be well-structured, it's always a good idea to keep an eye on performance considerations. Avoid unnecessary loops or nested functions that can impact the performance of the application.
   - Consider using caching mechanisms or optimizing database queries to improve the overall performance of the application.

4. Best practices:
   - Reusability is an important aspect of writing maintainable code. Consider extracting common functionalities into separate modules or utility functions that can be reused across different parts of the application.
   - Adhering to SOLID principles can help in writing more scalable and maintainable code. Make sure to follow principles like Single Responsibility and Dependency Injection to keep the codebase flexible and easy to extend.
   - Testability is crucial for ensuring the reliability of the code. Consider writing unit tests for critical components of the application to catch bugs early and facilitate future changes.

Overall, you are on the right track with your codebase. By incorporating these suggestions and best practices, you can further improve the quality and maintainability of the code. Keep up the good work!

File: src/libs/env/public.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    return publicEnvSchema.parse(env)
+  } catch (error) {
+    if (error instanceof z.ZodError) {
+      throw new Error(`환경 변수가 올바르게 설정되지 않았습니다. ${error.message}`)
+    }
+    throw error
+  }
+}
+
+export const publicEnv = validatePublicEnv()
```
Comment (PMtHk) [inline src/libs/env/public.ts:20]: Overall, this code snippet is well-organized and readable. The function `validatePublicEnv` is clear in its purpose of validating environment variables using a schema. The use of a try-catch block for error handling is a good practice to ensure that any potential errors are caught and handled appropriately.

1. Code organization & readability:
The naming of variables and functions is clear and descriptive, which is important for readability and understanding the code. The structure of the function is straightforward, making it easy to follow the logic.

Improvement:
To further improve code organization, consider breaking down the validation logic into smaller, more focused functions. This can help improve readability and maintainability by separating concerns.

2. Basic security & error handling:
The code includes basic error handling by catching any errors that occur during the validation process. This is important for ensuring that the application does not crash unexpectedly due to invalid environment variables.

Improvement:
Consider adding more specific error messages to provide better feedback to users or developers when environment variables are not set correctly. This can help in troubleshooting issues more effectively.

3. Performance considerations:
In terms of performance, the code snippet is efficient as it only performs the necessary validation of environment variables without any unnecessary operations.

4. Best practices:
The use of a schema for validating environment variables is a good practice as it helps ensure that the expected structure and types are enforced. This aligns with the SOLID principles by separating the validation logic from the rest of the code, making it easier to test and maintain.

Improvement:
Consider adding unit tests to validate the behavior of the `validatePublicEnv` function under different scenarios. This can help ensure that the function works as expected and catches any potential edge cases.

Overall, this code snippet demonstrates good practices in terms of code organization, error handling, and security. By following the suggested improvements and best practices, the code can be further enhanced in terms of readability, maintainability, and testability. Keep up the good work!

File: src/libs/env/schemas.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+  SUPABASE_URL: z.string().url(),
+  SUPABASE_ANON_KEY: z.string().min(1),
+})
+
+export const serverEnvSchema = z.object({
+  SUPABASE_SERVICE_ROLE_KEY: z.string().min(1),
+})
+
+export type PublicEnv = z.infer<typeof publicEnvSchema>
+export type ServerEnv = z.infer<typeof serverEnvSchema>
```
Comment (PMtHk) [inline src/libs/env/schemas.ts:13]: Overall, the code you have written is well-structured and readable. You have used a library like Zod for type validation, which is a good practice for ensuring data integrity. However, there are a few areas where improvements can be made:

1. Code organization & readability:
   - Naming: The naming of variables and functions is clear and descriptive, which is great for readability.
   - Structure: The separation of public and server environment schemas is a good way to organize the code.
   - Complexity: The code is simple and easy to follow, which is important for maintainability.

2. Basic security & error handling:
   - Security: You are validating the environment variables using Zod, which is a good practice to prevent injection attacks.
   - Error handling: It would be beneficial to add error handling for cases where the environment variables do not match the schema.

3. Performance considerations:
   - Performance: Since this code is mainly for validating environment variables, performance is not a major concern. However, it's always good to keep an eye on any potential bottlenecks as the codebase grows.

4. Best practices:
   - Reusability: You have created reusable schemas for public and server environments, which promotes code reusability.
   - SOLID principles: The code follows the Single Responsibility Principle by separating the validation logic into schemas.
   - Testability: The code can be easily tested by writing unit tests for the schema validation.

Example of improvement:
```typescript
import { z } from 'zod'

export const publicEnvSchema = z.object({
  SUPABASE_URL: z.string().url(),
  SUPABASE_ANON_KEY: z.string().min(1),
}).strict()

export const serverEnvSchema = z.object({
  SUPABASE_SERVICE_ROLE_KEY: z.string().min(1),
}).strict()

export type PublicEnv = z.infer<typeof publicEnvSchema>
export type ServerEnv = z.infer<typeof serverEnvSchema>
```

In the example above, I added the `.strict()` method to enforce strict mode in Zod, which helps catch any unexpected keys in the environment variables.

Overall, your code is well-structured and follows best practices. Keep up the good work and continue to focus on code quality and security.

File: src/libs/env/server.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+      throw new Error(`환경 변수가 올바르게 설정되지 않았습니다. ${error.message}`)
+    }
+
+    throw error
+  }
+}
+
+export async function getServerEnv(): Promise<ServerEnv> {
+  return await validateServerEnv()
+}
```
Comment (PMtHk) [inline src/libs/env/server.ts:29]: Overall, this code snippet is well-organized and readable, with clear naming conventions and a straightforward structure. However, there are a few areas where improvements can be made to enhance security, error handling, performance, and adherence to best practices.

1. Code organization & readability:
   - The use of `use server` at the beginning of the file is incorrect. It should be `use strict` for strict mode in JavaScript.
   - Consider using more descriptive variable names instead of abbreviations like `env`.
   - Add comments to explain the purpose of the functions and any complex logic.

Improvement example:
```javascript
'use strict';

import { z } from 'zod';
import { serverEnvSchema } from '@/libs/env/schemas';
import type { ServerEnv } from '@/libs/env/schemas';

// Validate the server environment configuration
async function validateServerEnv(): Promise<ServerEnv> {
  if (typeof window !== 'undefined') {
    throw new Error('This code cannot be executed in a client environment.');
  }

  try {
    const serverEnvConfig = {
      SUPABASE_SERVICE_ROLE_KEY: process.env.SUPABASE_SERVICE_ROLE_KEY,
    };

    return serverEnvSchema.parse(serverEnvConfig);
  } catch (error) {
    if (error instanceof z.ZodError) {
      throw new Error(`Environment variables are not correctly configured: ${error.message}`);
    }

    throw error;
  }
}

// Retrieve the validated server environment configuration
export async function getServerEnv(): Promise<ServerEnv> {
  return await validateServerEnv();
}
```

2. Basic security & error handling:
   - It's good to check for the environment before executing server-specific code to prevent potential security vulnerabilities.
   - Proper error handling ensures that any issues are caught and handled gracefully.

3. Performance considerations:
   - Since this code snippet is not computationally intensive, there are no significant performance considerations to address.

4. Best practices:
   - Consider adding unit tests to validate the behavior of the functions, promoting testability.
   - Ensure adherence to SOLID principles by separating concerns and maintaining a single responsibility for each function.

Overall, this code snippet demonstrates good practices in terms of organization and readability. By making the suggested improvements, you can further enhance the security, error handling, and maintainability of the codebase. Keep up the good work and continue to strive for clean, maintainable code.

File: src/utils/supabase/client.ts [hunk 5]
```diff
@@ -0,0 +1,8 @@
+import { createBrowserClient } from '@supabase/ssr'
+import { publicEnv } from '@/libs/env'
+
+import type { Database } from '@/utils/supabase/types'
+
+export function createClient() {
+  return createBrowserClient<Database>(publicEnv.SUPABASE_URL, publicEnv.SUPABASE_ANON_KEY)
+}
```
Comment (PMtHk) [inline src/utils/supabase/client.ts:8]: Overall, the code provided looks well-organized and readable. The naming conventions are clear, and the structure of the code is easy to follow. However, there are a few areas where improvements can be made:

1. Code organization & readability:
   - The function `createClient` is well-named and does what it says, which is great for readability.
   - Consider breaking down the import statements into separate lines for better readability.
   - It's a good practice to group related imports together for easier maintenance.

Improvement example:
```javascript
import { createBrowserClient } from '@supabase/ssr';
import { publicEnv } from '@/libs/env';
import type { Database } from '@/utils/supabase/types';

export function createClient() {
  return createBrowserClient<Database>(publicEnv.SUPABASE_URL, publicEnv.SUPABASE_ANON_KEY);
}
```

2. Basic security & error handling:
   - It's important to handle errors that may occur during the creation of the client, such as network issues or incorrect credentials.
   - Consider adding try-catch blocks to handle potential errors and provide meaningful error messages to the user.

3. Performance considerations:
   - The code snippet provided seems straightforward and should not have any significant performance issues.
   - However, it's always a good practice to review the performance impact of any external libraries or dependencies used in the code.

4. Best practices:
   - The code snippet follows the best practice of creating a separate function for creating a client, promoting reusability.
   - Consider implementing SOLID principles, such as Dependency Injection, to make the code more testable and maintainable.

In conclusion, the code provided is well-structured and readable, with room for minor improvements in error handling and import organization. By following best practices such as SOLID principles and ensuring proper error handling, the code can be further improved in terms of reusability, testability, and overall quality. Keep up the good work!

File: src/utils/supabase/server.ts [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+  const cookieStore = await cookies()
+
+  return createServerClient(publicEnv.SUPABASE_URL, serverEnv.SUPABASE_SERVICE_ROLE_KEY, {
+    cookies: {
+      getAll() {
+        return cookieStore.getAll()
+      },
+    },
+  })
+}
```
Comment (PMtHk) [inline src/utils/supabase/server.ts:18]: Overall, this code snippet is well-organized and readable. The naming of functions and variables is clear and descriptive, making it easy to understand the purpose of each part of the code. However, there are a few areas where improvements can be made:

1. Code organization & readability:
   - The import statements are not in a consistent order, which can make it harder to quickly scan and understand the dependencies of the module. It's a good practice to group imports by type (external, internal, local) and alphabetize them within each group.
   - The use of `await` outside of an async function (`use server`) is incorrect syntax and should be fixed. This can lead to runtime errors and should be avoided.

Improvement example:
```javascript
import { createServerClient } from '@supabase/ssr'
import { cookies } from 'next/headers'
import { publicEnv, getServerEnv } from '@/libs/env'

export async function createClient() {
  const serverEnv = await getServerEnv()
  const cookieStore = await cookies()

  return createServerClient(publicEnv.SUPABASE_URL, serverEnv.SUPABASE_SERVICE_ROLE_KEY, {
    cookies: {
      getAll() {
        return cookieStore.getAll()
      },
    },
  })
}
```

2. Basic security & error handling:
   - It's important to handle errors that may occur during the execution of asynchronous operations like fetching server environment variables or cookies. Adding try-catch blocks can help catch and handle these errors gracefully, preventing crashes and improving the overall robustness of the code.

3. Performance considerations:
   - The code snippet appears to be focused on setting up a client for server-side requests, which may not have significant performance considerations at this stage. However, as the application grows, it's important to keep an eye on potential bottlenecks and optimize where necessary.

4. Best practices:
   - The code snippet follows the best practice of separating concerns by creating a dedicated function (`createClient`) for setting up the server client. This promotes reusability and adheres to the Single Responsibility Principle (part of SOLID principles).
   - To improve testability, consider breaking down the logic into smaller, more testable functions and writing unit tests for each of them. This can help ensure that the code behaves as expected and is easier to maintain in the long run.

Overall, this code snippet is well-written and demonstrates good practices in terms of organization and readability. By addressing the minor issues mentioned above and continuing to follow best practices, the junior developer can further improve the quality of their code.

File: src/utils/supabase/types.ts [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+  CompositeTypeName extends PublicCompositeTypeNameOrOptions extends {
+    schema: keyof Database
+  }
+    ? keyof Database[PublicCompositeTypeNameOrOptions['schema']]['CompositeTypes']
+    : never = never,
+> = PublicCompositeTypeNameOrOptions extends { schema: keyof Database }
+  ? Database[PublicCompositeTypeNameOrOptions['schema']]['CompositeTypes'][CompositeTypeName]
+  : PublicCompositeTypeNameOrOptions extends keyof PublicSchema['CompositeTypes']
+    ? PublicSchema['CompositeTypes'][PublicCompositeTypeNameOrOptions]
+    : never
```
Comment (PMtHk) [inline src/utils/supabase/types.ts:269]: Overall, the code provided seems to be well-organized and readable. The use of TypeScript types and interfaces enhances code clarity and maintainability. However, there are some areas that could be improved to adhere to best practices and enhance the overall quality of the code.

1. Code organization & readability:
   - The code structure is clear and follows a consistent naming convention, which is beneficial for readability and maintainability.
   - Consider breaking down the large `Database` type into smaller, more manageable parts to improve readability and reduce complexity. This can be achieved by grouping related types together and using interfaces instead of nested objects.
   - Use meaningful names for types and variables to improve code understanding. For example, consider renaming `Json` to a more descriptive name like `JsonValue`.

   Example of improvement:
   ```typescript
   // Instead of a single large Database type, break it down into smaller interfaces
   interface PublicTables {
     categories: CategoryTable;
     posts: PostTable;
     // Add other tables here
   }

   interface CategoryTable {
     Row: {
       id: number;
       name: string;
     };
     Insert: {
       id?: number;
       name: string;
     };
     // Add other properties here
   }
   ```

   Relevant best practice: Use interfaces to define complex types and break down large types into smaller, more manageable parts.

2. Basic security & error handling:
   - Ensure that input validation is performed on user-provided data to prevent security vulnerabilities like SQL injection or data manipulation attacks.
   - Implement error handling mechanisms, such as try-catch blocks, to gracefully handle exceptions and prevent crashes in the application.

   Example of improvement:
   ```typescript
   function insertCategory(category: TablesInsert<'public', 'categories'>): void {
     // Validate input data before inserting into the database
     if (!category.name) {
       throw new Error('Category name is required');
     }
     // Insert category into the database
   }
   ```

   Relevant best practice: Implement input validation and error handling to enhance security and prevent unexpected behavior.

3. Performance considerations:
   - Consider optimizing database queries by using indexes, caching, or batch processing to improve performance.
   - Avoid unnecessary data fetching or processing to reduce latency and improve response times.

   Example of improvement:
   ```typescript
   // Use database indexes to speed up query performance
   CREATE INDEX idx_category_name ON categories (name);
   ```

   Relevant best practice: Optimize database queries and data processing to improve application performance.

4. Best practices:
   - Ensure adherence to SOLID principles by designing modular and maintainable code that is easy to extend and test.
   - Aim for reusability by creating generic types and functions that can be used across different parts of the application.

   Example of improvement:
   ```typescript
   // Create a generic function for updating database records
   function updateRecord<T>(tableName: string, data: T): void {
     // Update record in the database
   }
   ```

   Relevant best practice: Design code that is modular, reusable, and adheres to SOLID principles for better maintainability and testability.

Overall, the code shows promise and demonstrates a good understanding of TypeScript and code organization. By incorporating the suggested improvements and best practices, the code can be further enhanced in terms of readability, security, performance, and maintainability. Keep up the good work and continue learning and improving your coding skills!

**Review by PMtHk — COMMENTED** [review]
Code review by ChatGPT

File: src/utils/supabase/server.ts [hunk 8]
```diff
@@ -0,0 +1,18 @@
+'use server'
+
+import { createServerClient } from '@supabase/ssr'
+import { cookies } from 'next/headers'
+import { publicEnv, getServerEnv } from '@/libs/env'
+
+export async function createClient() {
+  const serverEnv = await getServerEnv()
+  const cookieStore = await cookies()
+
```
Comment (PMtHk) [inline src/utils/supabase/server.ts:10]: 코드 리뷰를 통해 다음과 같은 사항들을 개선할 수 있을 것으로 판단됩니다:

1. 코드 구성 및 가독성:
- 함수명과 변수명이 명확하고 의미 있는지 확인해야 합니다. 예를 들어, `createClient` 함수의 이름은 명확하지만 `serverEnv` 변수의 위치가 올바른지 확인이 필요합니다.
- 코드 구조가 명확하고 읽기 쉬운지 확인해야 합니다. 현재 코드에서는 `serverEnv` 변수가 함수 내부에 위치하고 있어 가독성이 떨어질 수 있습니다.

2. 기본 보안 및 오류 처리:
- 현재 코드에서는 오류 처리가 부족한 것으로 보입니다. 예외 상황에 대한 적절한 처리가 필요합니다. 예를 들어, `getServerEnv` 함수에서 발생할 수 있는 오류에 대한 처리가 필요합니다.

3. 성능 고려:
- 코드의 성능을 고려하여 최적화할 수 있는 부분이 있는지 확인해야 합니다. 현재 코드에서는 성능에 대한 고려가 부족한 것으로 보입니다.

4. 최선의 실천 방법:
- 코드의 재사용성, SOLID 원칙 및 테스트 가능성을 고려하여 개선할 수 있는 부분이 있는지 확인해야 합니다. 예를 들어, 코드를 모듈화하고 테스트 가능한 형태로 변경할 수 있습니다.

개선을 위해 코드를 다시 검토하고 위의 사항들을 고려하여 수정하는 것이 좋을 것으로 판단됩니다.

**Review by PMtHk — COMMENTED** [review]
Code review by ChatGPT
